src/burnchains/db.rs

```diff
@@ -147,16 +277,543 @@ impl<'a> BurnchainDBTransaction<'a> {
+                .expect("BUG: no metadata found for parent block-commit");
+
+        let (am, affirmed_reward_cycle) = if let Some(anchor_block) = anchor_block {
+            let anchor_am_id =
+                BurnchainDB::get_block_commit_affirmation_id(&self.sql_tx, &anchor_block)?
+                    .expect("BUG: anchor block has no affirmation map");
+
+            let mut am = BurnchainDB::get_affirmation_map(&self.sql_tx, anchor_am_id)?
+                .ok_or(BurnchainError::DBError(DBError::NotFoundError))?;
+
+            if descends_from_anchor_block {
+                test_debug!("Prepare-phase commit {},{},{} descends from anchor block {},{},{} for reward cycle {}",
+                            &block_commit.block_header_hash, block_commit.block_height, block_commit.vtxindex, &anchor_block.block_header_hash, anchor_block.block_height, anchor_block.vtxindex, reward_cycle);
+
+                am.push(AffirmationMapEntry::PoxAnchorBlockPresent);
+                (am, Some(reward_cycle))
+            } else {
+                test_debug!("Prepare-phase commit {},{},{} does NOT descend from anchor block {},{},{} for reward cycle {}",
+                            &block_commit.block_header_hash, block_commit.block_height, block_commit.vtxindex, &anchor_block.block_header_hash, anchor_block.block_height, anchor_block.vtxindex, reward_cycle);
+
+                am.push(AffirmationMapEntry::PoxAnchorBlockAbsent);
+                (am, parent_metadata.anchor_block_descendant)
+            }
+        } else {
+            let (parent_reward_cycle, _) =
+                get_parent_child_reward_cycles(&parent, block_commit, burnchain)
+                    .ok_or(BurnchainError::DBError(DBError::NotFoundError))?;
+
+            // load up the affirmation map for the last anchor block the parent affirmed
+            let (mut am, parent_rc_opt) = match parent_metadata.anchor_block_descendant {
+                Some(parent_ab_rc) => {
+                    // parent affirmed some past anchor block
+                    let (_, ab_metadata) = BurnchainDB::get_anchor_block_commit(&self.sql_tx, parent_ab_rc)?
+                            .expect(&format!("BUG: parent descends from a reward cycle with an anchor block ({}), but no anchor block found", parent_ab_rc));
+
+                    let mut am =
+                        BurnchainDB::get_affirmation_map(&self.sql_tx, ab_metadata.affirmation_id)?
+                            .expect("BUG: no affirmation map for parent commit's anchor block");
+
+                    test_debug!("Prepare-phase commit {},{},{} does nothing for reward cycle {}, but it builds on its parent which affirms anchor block for reward cycle {} ({}) (affirms? {})",
+                                    &block_commit.block_header_hash, block_commit.block_height, block_commit.vtxindex, reward_cycle, parent_ab_rc, &am, (am.len() as u64) < parent_ab_rc);
+
+                    if (am.len() as u64) < parent_ab_rc {
```

Comment:
When the parent block is not a prepare-phase block-commit.

This code is called for each prepare-phase block-commit to go and fill in the affirmation it made, now that we know whether or not a PoX anchor block exists.  If this condition is true, then the affirmation map we loaded for the parent does not yet reflect the decision miners made as to whether or not to confirm the last anchor block that this parent block descends from.  The resulting `.push()` fixes this -- since we're calculating `block_commit`'s affirmation map, we will construct it from the parent's affirmation map, plus the affirmation that was just made about this new anchor block.

This code does _not_ run if the `block_commit`'s parent is in the same prepare phase -- it's affirmation map will already reflect all reward cycles up to `parent_ab_rc` from a previous call to this method.